contracts/Bonuses.sol

```diff
@@ -0,0 +1,148 @@
+pragma solidity ^0.4.17;
+
+// ----------------------------------------------------------------------------
+// Processables
+//
+// Copyright (c) 2017 OpenST Ltd.
+// https://simpletoken.org/
+//
+// The MIT Licence.
+// ----------------------------------------------------------------------------
+
+import "./SimpleToken.sol";
+import "./SafeMath.sol";
+import "./Processables.sol";
+
+/// @title Bonuses
+/// @dev Aids transferring bonus tokens related to the token sale
+contract Bonuses is Processables {
+    using SafeMath for uint256;
+
+	// named "processables" to aid JS interaction with this and other contracts that inherit from Processables
+	mapping(address => Processable) public processables;
+
+	// Amount of gas sufficient to conduct a single process loop iteration and post-loop operations
+	uint256 public constant PROCESS_GAS_MINIMUM = 100000;
+
+	// Total bonuses amount remaining to process after locking
+	uint256 public remainingTotalBonuses = 0;
+
+	struct Processable {
+		uint256 amount;
+		bool processed;
+	}
+
+	SimpleToken public simpleToken;
+
+	event BonusAdded(address indexed _address, uint256 _amount);
+	event BonusProcessed(address indexed _address, uint256 _amount, uint256 _index);
+
+	// limits execution to SimpleToken.owner
+	modifier onlySTOwner() {
+        require(msg.sender == simpleToken.owner());
+        _;
+    }
+
+
+	function Bonuses(SimpleToken _simpleToken)
+			Processables()
+			public
+	{
+		require(address(_simpleToken) != address(0));
+		simpleToken = _simpleToken;
+	}
+
+	/// @dev Adds bonus information to processables
+	/// @param _address address
+	/// @param _amount amount
+	/// @return true
+	function add(address _address, uint256 _amount)
+			public
+			onlyOwner
+			onlyIfUnlocked
+			returns (bool result)
+	{
+		require(_address != address(0));
+		require(_amount > 0);
+
+		Processable storage processable = processables[_address];
+
+		require(processable.amount == 0);
+		require(add(_address));
+
+		processable.amount = _amount;
+
+		BonusAdded(_address, _amount);
+
+		return true;
+	}
+
+	/// @dev Locks this contract
+	function lock()
+			public
+			returns (bool result)
+	{
+		// onlyOwner
+		result = super.lock();
+
+		for (uint256 i = 0; i < addresses.length; i++)
```

Comment:
We'll just have to be careful not to add too many addresses else we won't be able to lock.

Reply:
As discussed, will move summing to `add` function. Was originally here due to an `update` function that has since been removed.